### PR TITLE
Feature: processes filter

### PR DESCRIPTION
**Pull request by lciolecki**
- FE: Refactor processes components
- BE: Add filtering by categories, isArchived, isSubprocess, isDeployed, processingTypes at /processes/ endpoint
- FE: Add filtering by categories, isSubprocess nad isDeployed
- FE: Add query params with filtering, pagination  and sorting

### CODE REVIEW

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/api/ProcessesResources.scala [hunk 1]
```diff
@@ -79,11 +80,25 @@ class ProcessesResources(val processRepository: FetchingProcessRepository,
...
+              'isSubprocess.as[Boolean].?,
+              'isArchived.as[Boolean].?,
+              'isDeployed.as[Boolean].?,
+              'categories.as(CsvSeq[String]).?,
+              'processingTypes.as(CsvSeq[String]).?
+            ) { (isSubprocess, isArchived, isDeployed, categories, processingTypes) =>
+              complete {
+                processRepository.fetchProcesses(
+                  isSubprocess,
+                  isArchived.orElse(Option(false)),
```
Comment (jedrz) [inline ui/server/src/main/scala/pl/touk/nussknacker/ui/api/ProcessesResources.scala]: Do we need orElse here?

File: ui/client/containers/Archive.js [hunk 2]
```diff
@@ -121,10 +141,9 @@ Archive.title = 'Archive Processes'
...
 Archive.header = 'Archive'
 
-function mapState(state) {
-  return {
-    loggedUser: state.settings.loggedUser
-  };
-}
+const mapState = (state) =>  ({
+  loggedUser: state.settings.loggedUser,
+  filterCategories: ProcessUtils.prepareFilterCategories(state.settings.loggedUser.categories, state.settings.loggedUser)
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/client/containers/Archive.js]: It's an back compability - there should by comment.

File: ui/client/common/ProcessUtils.js [hunk 3]
```diff
@@ -198,6 +198,31 @@ class ProcessUtils {
     }
   }
 
+  search = (processes, objectToFind) => {
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/client/common/ProcessUtils.js]: Hmm, mayby you'a right - I copy past this function..

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/api/ProcessesResources.scala (hunk 1 above)
Comment (lciolecki) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/api/ProcessesResources.scala]: It's an back compability - there should by comment.

File: ui/client/containers/Archive.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/client/containers/Archive.js]: I think so :)

File: ui/client/containers/BaseProcesses.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+  filterIsSubprocessOptions = [
+    {label: 'Show all types processes', value: null},
+    {label: 'Show only processes', value: false},
+    {label: 'Show only subprocesses', value: true},
+  ]
+
+  constructor(props) {
+    super(props)
+    this.query = queryString.parse(this.props.history.location.search, {arrayFormat: 'comma'})
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/client/containers/BaseProcesses.js]: Hmm.. Okej, I will change it.

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/FetchingProcessRepository.scala [hunk 5]
```diff
@@ -32,6 +32,9 @@ trait FetchingProcessRepository {
 
   def fetchProcesses()(implicit loggedUser: LoggedUser, ec: ExecutionContext): Future[List[ProcessDetails]]
 
+  def fetchProcesses(isSubprocess: Option[Boolean], isArchived: Option[Boolean], isDeployed: Option[Boolean], categories: Option[Seq[String]], processingTypes: Option[Seq[String]])
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/FetchingProcessRepository.scala]: This is good idea. But I will create ProcessQuery class when I will add Pagination, Sorting, Search on backend.

File: ui/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/DBFetchingProcessRepository.scala [hunk 6]
```diff
@@ -67,15 +84,23 @@ abstract class DBFetchingProcessRepository[F[_]](val dbConfig: DbConfig) extends
...
+
     (for {
       subprocessesVersions <- subprocessLastModificationDates
 
       tagsForProcesses <- tagsTable.result.map(_.toList.groupBy(_.processId).withDefaultValue(Nil))
 
-      latestProcesses <- processVersionsTable
+      deployedPerEnv <- latestDeployedProcessesVersionsPerEnvironment.result
+
+      latestProcesses <- processVersionsTable.filter{raw => if (isDeployed.isDefined) deployedPerEnv.map(_._1).contains(raw.id).equals(isDeployed.get):Rep[Boolean] else true:Rep[Boolean]} //TODO: move it to SLICK DSL
```
[the remark this answers is not in the document]
Comment (lciolecki) [reply ui/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/DBFetchingProcessRepository.scala]: Ok! I will change it.

File: ui/server/src/test/scala/pl/touk/nussknacker/ui/api/helpers/EspItTest.scala [hunk 7]
```diff
@@ -9,18 +9,29 @@ import cats.syntax.semigroup._
...
+import pl.touk.nussknacker.restmodel.process
+import pl.touk.nussknacker.ui.process.marshall.UiProcessMarshaller
 import pl.touk.nussknacker.ui.processreport.ProcessCounter
+import pl.touk.nussknacker.ui.security.api.{LoggedUser, Permission}
 
-trait EspItTest extends LazyLogging with WithHsqlDbTesting with TestPermissions { self: ScalatestRouteTest with Suite with BeforeAndAfterEach with Matchers =>
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+
+trait EspItTest extends LazyLogging with WithHsqlDbTesting with TestPermissions { self: ScalatestRouteTest with Suite with BeforeAndAfterEach with BeforeAndAfterAll with WithHsqlDbTesting with Matchers with ScalaFutures =>
```
Comment (mproch) [inline ui/server/src/test/scala/pl/touk/nussknacker/ui/api/helpers/EspItTest.scala]: Why you put WithHsqlDbTesting both in extends and in self?

File: ui/server/src/test/scala/pl/touk/nussknacker/ui/api/ProcessesResourcesSpec.scala [hunk 8]
```diff
@@ -160,6 +177,66 @@ class ProcessesResourcesSpec extends FunSuite with ScalatestRouteTest with Match
...
+    createProcess(secondProcessor, testCategoryName, false)
+    createDeployedProcess(thirdProcessor, testCategoryName, false)
+
+    Get(s"/processes") ~> routWithAllPermissions ~> check {
+      status shouldEqual StatusCodes.OK
+      val data = responseAs[String].decodeOption[List[BasicProcess]].get
+      data.size shouldBe 3
+    }
+
+    Get(s"/processes?isDeployed=true") ~> routWithAllPermissions ~> check {
```
Comment (mproch) [inline ui/server/src/test/scala/pl/touk/nussknacker/ui/api/ProcessesResourcesSpec.scala]: maybe we can add also assertion for ids ;)

File: ui/server/src/test/scala/pl/touk/nussknacker/ui/api/helpers/EspItTest.scala (hunk 7 above)
Comment (lciolecki) [reply ui/server/src/test/scala/pl/touk/nussknacker/ui/api/helpers/EspItTest.scala]: Ooow... This old bad changes. I will revert it.

File: ui/server/src/test/scala/pl/touk/nussknacker/ui/api/ProcessesResourcesSpec.scala (hunk 8 above)
Comment (lciolecki) [reply ui/server/src/test/scala/pl/touk/nussknacker/ui/api/ProcessesResourcesSpec.scala]: Good Idea :)

**Pull request merged**